exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/FragmentsRunner.java

```diff
@@ -0,0 +1,439 @@
+import org.apache.drill.exec.work.EndpointListener;
+import org.apache.drill.exec.work.WorkManager.WorkerBee;
+import org.apache.drill.exec.work.fragment.FragmentExecutor;
+import org.apache.drill.exec.work.fragment.FragmentStatusReporter;
+import org.apache.drill.exec.work.fragment.NonRootFragmentManager;
+import org.apache.drill.exec.work.fragment.RootFragmentManager;
+
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Is responsible for submitting query fragments for running (locally and remotely).
+ */
+public class FragmentsRunner {
+
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(FragmentsRunner.class);
+  private static final ControlsInjector injector = ControlsInjectorFactory.getInjector(FragmentsRunner.class);
+
+  private static final long RPC_WAIT_IN_MSECS_PER_FRAGMENT = 5000;
+
+  private final WorkerBee bee;
+  private final UserClientConnection initiatingClient;
+  private final DrillbitContext drillbitContext;
+  private final Foreman foreman;
+
+  private List<PlanFragment> planFragments;
+  private PlanFragment rootPlanFragment;
+  private FragmentRoot rootOperator;
+
+  public FragmentsRunner(WorkerBee bee, UserClientConnection initiatingClient, DrillbitContext drillbitContext, Foreman foreman) {
+    this.bee = bee;
+    this.initiatingClient = initiatingClient;
+    this.drillbitContext = drillbitContext;
+    this.foreman = foreman;
+  }
+
+  public WorkerBee getBee() {
+    return bee;
+  }
+
+  public void setPlanFragments(List<PlanFragment> planFragments) {
+    this.planFragments = planFragments;
+  }
+
+  public void setRootPlanFragment(PlanFragment rootPlanFragment) {
+    this.rootPlanFragment = rootPlanFragment;
+  }
+
+  public void setRootOperator(FragmentRoot rootOperator) {
+    this.rootOperator = rootOperator;
+  }
+
+  /**
+   * Submits root and non-root fragments fragments for running.
+   * In case of success move query to the running state.
+   */
+  public void submit() {
+    try {
+      assert planFragments != null;
+      assert rootPlanFragment != null;
+      assert rootOperator != null;
+
+      QueryId queryId = foreman.getQueryId();
+      assert queryId == rootPlanFragment.getHandle().getQueryId();
+
+      QueryManager queryManager = foreman.getQueryManager();
+
+      try {
+        drillbitContext.getWorkBus().addFragmentStatusListener(queryId, queryManager.getFragmentStatusListener());
+        drillbitContext.getClusterCoordinator().addDrillbitStatusListener(queryManager.getDrillbitStatusListener());
+
+        logger.debug("Submitting fragments to run.");
+        // set up the root fragment first so we'll have incoming buffers available.
+        setupRootFragment(rootPlanFragment, rootOperator);
+        setupNonRootFragments(planFragments);
+
+      } catch (ExecutionSetupException e) {
```

Comment:
We catch only `ExceptionSetupException`. But, Drill loves to throw unchecked exceptions. Should we handle those as well? Or, are they handled elsewhere? If elsewhere, why do we handle this one exception here?